File: backend/services/matching/matching/domain/pricing.py
"""Port of ``src/domain/pricing.ts``."""

from __future__ import annotations

from cappy_common.jsmath import js_round
from cappy_common.models import AnyListing, AnyRequirement, Cents, Quote

from .money import bps

# 15% take rate. The fee sits inside the total the buyer pays: it is not added on
# top at the last step, because that is the thing everyone hates about marketplaces.
PLATFORM_FEE_BPS = 1500


def hours_for(req: AnyRequirement, listing: AnyListing) -> float | None:
    """How many hours of the asset this request consumes.

    window: the buyer names the duration.
    batch:  setup plus run time, derived from quantity and throughput.
    None when the request and listing are different shapes; feasibility rejects those.
    """
    if req.mode == "window" and listing.mode == "window":
        return req.hours
    if req.mode == "batch" and listing.mode == "batch":
        return listing.setup_hours + req.quantity / listing.units_per_hour
    return None


def quote_for(req: AnyRequirement, listing: AnyListing) -> Quote | None:
    hours = hours_for(req, listing)
    if hours is None:
        return None

    base: Cents = js_round(hours * listing.rate_per_hour)
    if listing.mode == "window":
        extra, extra_label = listing.extra_fee, listing.extra_label
    else:
        extra, extra_label = listing.setup_fee, "Setup and programming"
    total: Cents = base + extra
    platform_fee = bps(total, PLATFORM_FEE_BPS)

    return Quote(
        hours=hours,
        base=base,
        extra=extra,
        extra_label=extra_label,
        total=total,
        platform_fee=platform_fee,
        owner_net=total - platform_fee,
    )

File: cappy/cappy/src/domain/pricing.ts
import type { Cents, Listing, Quote, Requirement } from './types.ts'
import { isBatch, isWindow } from './types.ts'
import { bps } from './money.ts'

/** 15% take rate. The fee sits inside the total the buyer pays, it is not added on
 *  top at the last step, because that is the thing everyone hates about marketplaces. */
export const PLATFORM_FEE_BPS = 1500

/**
 * How many hours of the asset this request consumes.
 * `window`, the buyer names the duration.
 * `batch` , setup plus run time, derived from quantity and throughput.
 * Null when the request and listing are different shapes; feasibility rejects those.
 */
export function hoursFor(req: Requirement, listing: Listing): number | null {
  if (req.mode === 'window' && isWindow(listing)) return req.hours
  if (req.mode === 'batch' && isBatch(listing)) {
    return listing.setupHours + req.quantity / listing.unitsPerHour
  }
  return null
}

export function quoteFor(req: Requirement, listing: Listing): Quote | null {
  const hours = hoursFor(req, listing)
  if (hours === null) return null

  const base: Cents = Math.round(hours * listing.ratePerHour)
  const extra: Cents = isWindow(listing) ? listing.extraFee : listing.setupFee
  const extraLabel = isWindow(listing) ? listing.extraLabel : 'Setup and programming'
  const total: Cents = base + extra
  const platformFee = bps(total, PLATFORM_FEE_BPS)

  return {
    hours,
    base,
    extra,
    extraLabel,
    total,
    platformFee,
    ownerNet: total - platformFee,
  }
}
